File: 01_Array/Q12_380. Insert Delete GetRandom O(1)/solution.cpp
class RandomizedSet {
public:

    vector<int> vec;
    unordered_map<int, int> mp;

    RandomizedSet() {
        
    }
    
    bool insert(int val) {
        if(mp.find(val) != mp.end()) return false;

        vec.push_back(val);
        mp[val] = vec.size() - 1;
        return true;
    }
    
    bool remove(int val) {
        if(mp.find(val) == mp.end()) return false;

        int idx = mp[val];
        int lastElement = vec.back();
        vec[idx] = lastElement;
        mp[lastElement] = idx;
        vec.pop_back();
        mp.erase(val);
        return true;
    }
    
    int getRandom() {
        int randomIdx = rand() % vec.size();
        return vec[randomIdx];
    }
};



// Time Complexity -> O(1)
// Space Complexity ->O(n)

File: 01_Array/Q12_380. Insert Delete GetRandom O(1)/solution.py
class RandomizedSet:

    def __init__(self):
        self.numList = []
        self.numMap = {}

    def insert(self, val: int) -> bool:
        res = val not in self.numMap
        if res:
            self.numMap[val] = len(self.numList)
            self.numList.append(val)
        return res


    def remove(self, val: int) -> bool:
        res = val in self.numMap
        if res:
            idx = self.numMap[val]
            lastValue = self.numList[-1]
            self.numList[idx] = lastValue
            self.numList.pop()
            self.numMap[lastValue] = idx
            del self.numMap[val]
        return res

    def getRandom(self) -> int:
        return random.choice(self.numList)


# Time Complexity -> O(1)
# Space Complexity -> O(n)
